File: scripts/scraper.ts
/**
 * JaveaHomeFinders.com Property Scraper
 *
 * Scrapes property listings from javeahomefinders.com and converts them
 * to the Miraluna property format.
 *
 * Usage:
 *   npx tsx scripts/scraper.ts
 */

import * as cheerio from 'cheerio';

// ============================================================================
// TYPES
// ============================================================================

export interface ScrapedProperty {
  // Basic Info
  reference: string;
  title: string;
  price: number;
  currency: string;

  // Property Details
  type: 'villa' | 'apartment' | 'townhouse' | 'plot' | 'investment';
  location: string;
  bedrooms: number;
  bathrooms: number;
  buildSize: number;
  plotSize?: number;

  // Features
  hasPool: boolean;
  features: string[];

  // Description
  description: string;
  descriptionSpanish?: string; // Will be translated

  // Images
  images: string[];
  mainImage: string;

  // Metadata
  sourceUrl: string;
  scrapedAt: Date;
  status?: 'available' | 'reserved' | 'sold';
  tags?: string[];
}

// ============================================================================
// SCRAPER FUNCTIONS
// ============================================================================

/**
 * Fetches HTML content from a URL
 */
async function fetchHTML(url: string): Promise<string> {
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to fetch ${url}: ${response.statusText}`);
  }
  return await response.text();
}

/**
 * Scrapes the property listing page to get all property URLs
 */
export async function scrapePropertyList(baseUrl: string = 'https://www.javeahomefinders.com'): Promise<string[]> {
  const searchUrl = `${baseUrl}/search-property?trans=sale&city=85182`; // Jávea properties
  const html = await fetchHTML(searchUrl);
  const $ = cheerio.load(html);

  const propertyUrls: string[] = [];

  // Find all property cards
  $('.featDetailCont a').each((_, element) => {
    const href = $(element).attr('href');
    if (href && !href.includes('javascript')) {
      const fullUrl = href.startsWith('http') ? href : `${baseUrl}${href}`;
      propertyUrls.push(fullUrl);
    }
  });

  // Handle pagination - get all pages
  const totalPages = $('.pagBtn').length;
  for (let page = 2; page <= totalPages; page++) {
    const pageUrl = `${searchUrl}&page=${page}`;
    const pageHtml = await fetchHTML(pageUrl);
    const $page = cheerio.load(pageHtml);

    $page('.featDetailCont a').each((_, element) => {
      const href = $page(element).attr('href');
      if (href && !href.includes('javascript')) {
        const fullUrl = href.startsWith('http') ? href : `${baseUrl}${href}`;
        if (!propertyUrls.includes(fullUrl)) {
          propertyUrls.push(fullUrl);
        }
      }
    });
  }

  return [...new Set(propertyUrls)]; // Remove duplicates
}

/**
 * Extracts text content and cleans it
 */
function cleanText(text: string | undefined): string {
  return (text || '').trim().replace(/\s+/g, ' ');
}

/**
 * Extracts number from text (e.g., "€320,000" => 320000)
 */
function extractNumber(text: string): number {
  const match = text.replace(/[€,\s]/g, '').match(/\d+/);
  return match ? parseInt(match[0], 10) : 0;
}

/**
 * Maps property type from source to our format
 */
function mapPropertyType(typeText: string): ScrapedProperty['type'] {
  const lower = typeText.toLowerCase();
  if (lower.includes('villa')) return 'villa';
  if (lower.includes('apartment') || lower.includes('penthouse')) return 'apartment';
  if (lower.includes('townhouse')) return 'townhouse';
  if (lower.includes('plot') || lower.includes('land')) return 'plot';
  return 'apartment'; // default
}

/**
 * Scrapes a single property detail page
 */
export async function scrapePropertyDetail(url: string): Promise<ScrapedProperty> {
  const html = await fetchHTML(url);
  const $ = cheerio.load(html);

  // Extract reference number
  const refText = $('.propRefCont span').first().text();
  const reference = refText.replace('Ref:', '').trim() || url.split('/').pop() || 'unknown';

  // Extract title
  const title = cleanText($('#propertyTitle h1').text()) || 'Untitled Property';

  // Extract price
  const priceText = $('.propRefCont .pricePV').text();
  const price = extractNumber(priceText);
  const currency = priceText.includes('€') ? 'EUR' : 'USD';

  // Extract type and location
  const typeText = $('.bulletList li').first().text();
  const locationText = $('.bulletList li').eq(1).text();

  const type = mapPropertyType(typeText);
  const location = locationText.replace(/.*:/, '').trim();

  // Extract bedrooms, bathrooms
  let bedrooms = 0;
  let bathrooms = 0;
  $('.bulletList li').each((_, el) => {
    const text = $(el).text();
    if (text.includes('Bedroom')) {
      bedrooms = extractNumber(text);
    }
    if (text.includes('Bathroom')) {
      bathrooms = extractNumber(text);
    }
  });

  // Extract build and plot size
  let buildSize = 0;
  let plotSize: number | undefined;
  $('.bulletList li').each((_, el) => {
    const text = $(el).text();
    if (text.includes('Build')) {
      buildSize = extractNumber(text);
    }
    if (text.includes('Terrace Size') || text.includes('Plot')) {
      plotSize = extractNumber(text);
    }
  });

  // Check for pool
  const hasPool = $('.bulletList li').text().toLowerCase().includes('pool') ||
                   $('.fa-swimming-pool').length > 0;

  // Extract features
  const features: string[] = [];
  $('.bulletList li .fa-check').each((_, el) => {
    const featureText = $(el).parent().text().trim();
    if (featureText) {
      features.push(featureText);
    }
  });

  // Extract description
  const description = cleanText($('.propDescCont').text());

  // Extract images
  const images: string[] = [];
  $('#mainPhotos li').each((_, el) => {
    const bgImage = $(el).find('.mainPhotoImgContainer a').css('background-image');
    if (bgImage) {
      const match = bgImage.match(/url\(['"]?([^'"]+)['"]?\)/);
      if (match) {
        images.push(match[1]);
      }
    }
  });

  const mainImage = images[0] || '';

  // Extract status from overlay
  let status: ScrapedProperty['status'] = 'available';
  const overlayText = $('.photoOverlayText').text().toLowerCase();
  if (overlayText.includes('reserved')) {
    status = 'reserved';
  } else if (overlayText.includes('sold')) {
    status = 'sold';
  }

  // Extract tags
  const tags: string[] = [];
  if (overlayText.includes('new build')) tags.push('new-build');
  if (overlayText.includes('luxury')) tags.push('luxury');
  if (overlayText.includes('sea view')) tags.push('sea-view');

  return {
    reference,
    title,
    price,
    currency,
    type,
    location,
    bedrooms,
    bathrooms,
    buildSize,
    plotSize,
    hasPool,
    features,
    description,
    images,
    mainImage,
    sourceUrl: url,
    scrapedAt: new Date(),
    status,
    tags,
  };
}

/**
 * Scrapes all properties from javeahomefinders.com
 */
export async function scrapeAllProperties(): Promise<ScrapedProperty[]> {
  console.log('🔍 Finding property URLs...');
  const urls = await scrapePropertyList();
  console.log(`📋 Found ${urls.length} properties to scrape`);

  const properties: ScrapedProperty[] = [];

  for (let i = 0; i < urls.length; i++) {
    try {
      console.log(`\n[${i + 1}/${urls.length}] Scraping: ${urls[i]}`);
      const property = await scrapePropertyDetail(urls[i]);
      properties.push(property);
      console.log(`✅ ${property.reference} - ${property.title}`);

      // Be polite - wait 1 second between requests
      if (i < urls.length - 1) {
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    } catch (error) {
      console.error(`❌ Error scraping ${urls[i]}:`, error);
    }
  }

  console.log(`\n✨ Scraping complete! Scraped ${properties.length} properties`);
  return properties;
}

// ============================================================================
// CLI EXECUTION
// ============================================================================

if (require.main === module) {
  scrapeAllProperties()
    .then(properties => {
      console.log('\n📊 Scraping Results:');
      console.log(`Total properties: ${properties.length}`);
      console.log(`Villas: ${properties.filter(p => p.type === 'villa').length}`);
      console.log(`Apartments: ${properties.filter(p => p.type === 'apartment').length}`);
      console.log(`Townhouses: ${properties.filter(p => p.type === 'townhouse').length}`);
      console.log(`Plots: ${properties.filter(p => p.type === 'plot').length}`);

      // Save to JSON file
      const fs = require('fs');
      const path = require('path');
      const outputPath = path.join(__dirname, '..', 'scraped-properties.json');
      fs.writeFileSync(outputPath, JSON.stringify(properties, null, 2));
      console.log(`\n💾 Saved to: ${outputPath}`);
    })
    .catch(error => {
      console.error('Fatal error:', error);
      process.exit(1);
    });
}

File: scripts/scraper.py
#!/usr/bin/env python3
"""
JaveaHomeFinders Property Scraper (Python)

Scrapes property listings from javeahomefinders.com and uploads to Supabase.

Requirements:
    pip install requests beautifulsoup4 python-dotenv supabase

Usage:
    python scripts/scraper.py
"""

import os
import json
import re
import time
from datetime import datetime
from typing import List, Dict, Optional
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup
from dotenv import load_dotenv
from supabase import create_client, Client

# Load environment variables from .env.local
load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_SERVICE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')

# ============================================================================
# SCRAPER FUNCTIONS
# ============================================================================

def fetch_html(url: str) -> str:
    """Fetch HTML content from a URL."""
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
    }
    response = requests.get(url, headers=headers)
    response.raise_for_status()
    return response.text


def scrape_property_list(base_url: str = 'https://www.javeahomefinders.com') -> List[str]:
    """Scrape the property listing page to get all property URLs."""
    search_url = f"{base_url}/search-property?trans=sale&city=85182"  # Jávea properties
    html = fetch_html(search_url)
    soup = BeautifulSoup(html, 'html.parser')

    property_urls = []

    # Find all property cards
    for link in soup.select('.featDetailCont a'):
        href = link.get('href')
        if href and 'javascript' not in href:
            full_url = href if href.startswith('http') else urljoin(base_url, href)
            if full_url not in property_urls:
                property_urls.append(full_url)

    # Handle pagination
    pagination = soup.select('.pagBtn')
    total_pages = len(pagination) if pagination else 1

    for page in range(2, total_pages + 1):
        page_url = f"{search_url}&page={page}"
        page_html = fetch_html(page_url)
        page_soup = BeautifulSoup(page_html, 'html.parser')

        for link in page_soup.select('.featDetailCont a'):
            href = link.get('href')
            if href and 'javascript' not in href:
                full_url = href if href.startswith('http') else urljoin(base_url, href)
                if full_url not in property_urls:
                    property_urls.append(full_url)

    return list(set(property_urls))  # Remove duplicates


def extract_number(text: str) -> int:
    """Extract number from text (e.g., '€320,000' => 320000)."""
    match = re.search(r'\d+', text.replace(',', '').replace('€', '').replace(' ', ''))
    return int(match.group()) if match else 0


def normalize_location_text(text: str) -> str:
    """Normalize location text to handle Valencian/Spanish variations.

    Xàbia (Valencian) = Javea (Spanish)
    """
    # Replace Valencian spelling with Spanish
    text = re.sub(r'X[àa]bia', 'Javea', text, flags=re.IGNORECASE)
    return text


def map_property_type(type_text: str) -> str:
    """Map property type from source to our format."""
    lower = type_text.lower()

    # Apartment/Penthouse
    if 'apartment' in lower or 'penthouse' in lower:
        return 'apartment'

    # Villa/House
    if 'villa' in lower:
        return 'house'

    # Townhouse
    if 'townhouse' in lower or 'town house' in lower:
        return 'house'

    # Plot/Land
    if 'plot' in lower or 'land' in lower:
        return 'plot'

    # Default
    return 'house'


def map_location_to_area(location: str) -> str:
    """Map location to area code."""
    lower = location.lower()

    # Jávea areas
    if 'old town' in lower or 'pueblo' in lower:
        return 'old-town'
    if 'arenal' in lower or 'beach' in lower:
        return 'arenal'
    if 'port' in lower or 'puerto' in lower:
        return 'port'
    if 'cap marti' in lower:
        return 'cap-marti'
    if 'granadella' in lower:
        return 'granadella'
    if 'montgo' in lower or 'montgó' in lower:
        return 'montgo'
    if 'pinosol' in lower:
        return 'pinosol'

    # Other locations
    if 'moraira' in lower:
        return 'moraira'
    if 'denia' in lower or 'dénia' in lower:
        return 'denia'
    if 'benitachell' in lower:
        return 'benitachell'

    return 'other'


def generate_property_id(reference: str) -> str:
    """Generate a clean ID from reference."""
    return re.sub(r'[^a-z0-9]+', '-', reference.lower()).strip('-')


def scrape_property_detail(url: str) -> Dict:
    """Scrape a single property detail page."""
    html = fetch_html(url)
    soup = BeautifulSoup(html, 'html.parser')

    # Extract reference
    ref_elem = soup.select_one('.propRefCont span')
    reference = ref_elem.text.replace('Ref:', '').strip() if ref_elem else url.split('/')[-1]

    # Extract title
    title_elem = soup.select_one('#propertyTitle h1')
    title = title_elem.text.strip() if title_elem else 'Untitled Property'
    title = normalize_location_text(title)  # Normalize Xàbia → Javea in title

    # Extract price
    price_elem = soup.select_one('.propRefCont .pricePV')
    price_text = price_elem.text if price_elem else '0'
    price = extract_number(price_text)

    # Extract type and location
    bullet_items = soup.select('.bulletList li')
    type_text = bullet_items[0].text if len(bullet_items) > 0 else ''
    location_text = bullet_items[1].text if len(bullet_items) > 1 else ''

    property_type = map_property_type(type_text)

    # Parse location: format is like "Jávea, Pueblo" or "Xàbia, Arenal"
    # Remove any icons/labels and normalize
    location = re.sub(r'.*:', '', location_text).strip()
    location = normalize_location_text(location)  # Xàbia → Javea

    # Extract bedrooms, bathrooms, sizes
    bedrooms = 0
    bathrooms = 0
    build_size = 0
    plot_size = None

    for item in bullet_items:
        text = item.text
        if 'Bedroom' in text:
            bedrooms = extract_number(text)
        if 'Bathroom' in text:
            bathrooms = extract_number(text)
        if 'Build' in text:
            build_size = extract_number(text)
        if 'Terrace Size' in text or 'Plot' in text:
            plot_size = extract_number(text)

    # Check for pool
    has_pool = 'pool' in soup.text.lower() or soup.select_one('.fa-swimming-pool') is not None

    # Extract features
    features = []
    if has_pool:
        features.append('pool')
    if plot_size and plot_size > 0:
        features.append('garden')

    for item in soup.select('.bulletList li .fa-check'):
        feature_text = item.parent.text.strip()
        if feature_text:
            if 'parking' in feature_text.lower():
                features.append('parking')
            if 'terrace' in feature_text.lower():
                features.append('terrace')
            if 'balcony' in feature_text.lower():
                features.append('balcony')
            if 'view' in feature_text.lower():
                features.append('views')
            if 'air con' in feature_text.lower():
                features.append('air-conditioning')
            if 'storage' in feature_text.lower():
                features.append('storage')

    features = list(set(features))  # Remove duplicates

    # Extract description
    desc_elem = soup.select_one('.propDescCont')
    description = desc_elem.text.strip() if desc_elem else ''
    description = normalize_location_text(description)  # Normalize Xàbia → Javea in description

    # Extract images
    images = []
    for photo in soup.select('#mainPhotos li'):
        img_container = photo.select_one('.mainPhotoImgContainer a')
        if img_container:
            style = img_container.get('style', '')
            match = re.search(r'url\(["\']?([^"\']+)["\']?\)', style)
            if match:
                images.append(match.group(1))

    # Extract status
    status = 'available'
    overlay = soup.select_one('.photoOverlayText')
    if overlay:
        overlay_text = overlay.text.lower()
        if 'reserved' in overlay_text:
            status = 'reserved'
        elif 'sold' in overlay_text:
            status = 'sold'

    return {
        'id': generate_property_id(reference),
        'type': property_type,
        'title': title,
        'price': price,
        'location': location,
        'description': description,
        'images': images,
        'features': features,
        'specs': {
            'bedrooms': bedrooms if bedrooms > 0 else None,
            'bathrooms': bathrooms if bathrooms > 0 else None,
            'size': build_size if build_size > 0 else None,
            'plotSize': plot_size if plot_size and plot_size > 0 else None,
        },
        'source_url': url,
        'source_reference': reference,
        'status': status,
        'scraped_at': datetime.now().isoformat(),
    }


def scrape_all_properties() -> List[Dict]:
    """Scrape all properties from javeahomefinders.com."""
    print('🔍 Finding property URLs...')
    urls = scrape_property_list()
    print(f'📋 Found {len(urls)} properties to scrape')

    properties = []

    for i, url in enumerate(urls):
        try:
            print(f'\n[{i + 1}/{len(urls)}] Scraping: {url}')
            property_data = scrape_property_detail(url)
            properties.append(property_data)
            print(f'✅ {property_data["id"]} - {property_data["title"]}')

            # Be polite - wait 1 second between requests
            if i < len(urls) - 1:
                time.sleep(1)
        except Exception as e:
            print(f'❌ Error scraping {url}: {e}')

    print(f'\n✨ Scraping complete! Scraped {len(properties)} properties')
    return properties


# ============================================================================
# DATABASE FUNCTIONS
# ============================================================================

def get_supabase_client() -> Client:
    """Create and return a Supabase client."""
    if not SUPABASE_URL or not SUPABASE_SERVICE_KEY:
        raise ValueError(
            'Missing Supabase credentials. Add SUPABASE_SERVICE_ROLE_KEY to .env.local'
        )
    return create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)


def upload_properties_to_supabase(properties: List[Dict]) -> None:
    """Upload properties to Supabase database."""
    supabase = get_supabase_client()

    print(f'\n📤 Uploading {len(properties)} properties to Supabase...')

    success_count = 0
    error_count = 0

    for prop in properties:
        try:
            # Upsert (insert or update if exists)
            result = supabase.table('properties').upsert(prop).execute()
            print(f'✅ Uploaded: {prop["id"]}')
            success_count += 1
        except Exception as e:
            print(f'❌ Error uploading {prop["id"]}: {e}')
            error_count += 1

    print(f'\n📊 Upload Summary:')
    print(f'✅ Success: {success_count}')
    print(f'❌ Errors: {error_count}')


def save_to_json(properties: List[Dict], filename: str = 'scraped-properties.json') -> None:
    """Save properties to a JSON file."""
    with open(filename, 'w', encoding='utf-8') as f:
        json.dump(properties, f, indent=2, ensure_ascii=False)
    print(f'\n💾 Saved to: {filename}')


# ============================================================================
# MAIN EXECUTION
# ============================================================================

def main():
    """Main execution function."""
    import sys

    print('🕷️  JaveaHomeFinders Property Scraper\n')

    # Check for --upload flag (default is NO upload)
    should_upload = '--upload' in sys.argv

    if should_upload:
        print('📤 Upload to Supabase: ENABLED')
    else:
        print('📤 Upload to Supabase: DISABLED (use --upload to enable)')
    print()

    # Scrape properties
    properties = scrape_all_properties()

    # Save to JSON file (backup)
    save_to_json(properties)

    # Upload to Supabase (only if --upload flag is set)
    if should_upload:
        try:
            upload_properties_to_supabase(properties)
        except ValueError as e:
            print(f'\n⚠️  {e}')
            print('Properties saved to JSON file only.')
    else:
        print('\n✨ Scraping complete! (Skipped upload - use --upload to enable)')


if __name__ == '__main__':
    main()
